pallets/stateful-storage/src/stateful_child_tree.rs

```diff
@@ -0,0 +1,78 @@
+use codec::{Decode, Encode};
+use common_primitives::msa::MessageSourceId;
+use frame_support::{
+	storage::{child, child::ChildInfo, ChildTriePrefixIterator},
+	Identity, StorageHasher,
+};
+use sp_core::hexdisplay::AsBytesRef;
+use sp_std::prelude::*;
+
+pub type StatefulPageKeyPart = Vec<u8>;
+
+/// Paginated Stateful data access utility
+pub struct StatefulStorageTree;
+impl StatefulStorageTree {
+	pub fn concat_keys(keys: &[StatefulPageKeyPart]) -> Vec<u8> {
+		let mut key = Vec::<u8>::new();
+		for k in keys {
+			let mut k2 = k.to_owned();
+			key.append(&mut k2);
+		}
+
+		key
+	}
+
+	/// Reads a child tree node
+	///
+	/// The read is performed from the `msa_id` only. The `key` is not required. If the
+	/// data doesn't store under the given `key` `None` is returned.
+	pub fn read<V: Decode + Sized>(
```

Comment:
I improved this read to be able to differentiate between empty storage with decoding issue. Please replace this with  https://github.com/LibertyDSNP/frequency/pull/1017/commits/2289e8d811ed9c05560cc574ee57941dab60a369#diff-e3a606fd6c211ed528c4fb3f8f62a5be19573b974946cdf75a0c3a173e143fd3R16-R26